### ISSUE TITLE
StringBuilder should have all functionality of String

### ISSUE THREAD

**Issue by int19h** [dotnet/corefx#11390]
`StringBuilder` is, essentially, just a mutable version of `String`. Consequently, it would make sense for it to have all the functionality that `String` has, with in-place mutating versions where appropriate. Right now it has a subset that doesn't seem particularly coherent - e.g. there's indexing, but no `Substring`. Some other missing members include `IndexOf`, `Contains`, `Find`, `Pad...`, `ToLower` and `ToUpper` etc. Basically, it should be possible to do anything you could do with a `String` without actually creating a string (to avoid unnecessary copies).

There is also a matter of consistency - the same pattern with an immutable value class and a mutable builder version is also repeated with `Uri` and `UriBuilder`, but there the builder does actually reproduce the API surface of the immutable class much closer.

**Comment by vancem** [comment]
I agree with this basic notion that StringBuilder should be basically an 'advanced' string.    Certainly having some operations like substring and indexOf make sense.    

I also have a desire to see if we can't make StringBuilder work for all the scenarios where various special-purpose strings have been invented.  This includes mutable strings, but it also includes large strings (where you want constant time concatenation).   I think some sort of copy-on-write reference mechanism can work for this and thus we can get down to just two strings representations.  

It is likely that the two issues (adding string APIs to StringBuilder), and allowing it to work efficiently in other scenarios (large or mutable strings), are largely independent of one another, so we could do the first step (add the APIs) before the second (change the implementation to allow efficiency in the other scenarios).   Still I would want these second cases thought out before we commit to the first phase.  

The first step is reasonably straightforward (a pull request with implementations for the new APIs), if anyone wants to step up and start the ball rolling.

**Comment by svick** [comment]
@vancem 

> The first step is reasonably straightforward (a pull request with implementations for the new APIs)

Shouldn't it follow [the API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md)? I. e. a "speclet" first, PR only after the API is reviewed.

**Comment by vancem** [comment]
Yes, we will follow the API review process.

**Comment by danmosemsft** [comment]
@int19h to be concrete, we invite a speclet following the API review process. when that looks in good shape, we can take to API review.

**Comment by int19h** [comment]
Is there a need for anything other than the list of desired members? It seems that all the use cases would be identical to those for equivalent members on regular strings, for example.

**Comment by vancem** [comment]
Normally you create the reference assembly for the new API (thus all method bodies stubbed out to do nothing).  

There are interesting design issues, like does 'SubString' return a String or a StringBuilder?     Also which APIs would be in place vs returning a new value.    Another interesting question is, given that it is not going to be a 'Drop in' replacement (since some things will be 'in place'), how important is parity with String?  As long as the API is 'complete' the rest may not be worth it.  

But perhaps most importantly, are compelling examples of the performance advantage of having such APIs (that is places where a lot of copies are avoided).  

The basic idea is interesting, we have to figure out how much merit it really has...

**Comment by AlexGhiondea** [comment]
It sounds like there are some challenging design questions that need to be answered before we move forward with the implementation.

Anyone interested in thinking about this and coming up with a proposal?

**Comment by magol** [comment]
This issue have a similar questions as "String-like extension methods to ReadOnlySpan<char> Epic" (#21395). 

The solutions in that issue was to return a ReadOnlySpan<char> (when appropriate), and I think that we can use the same philosophy here (that we should return a StringBuilder)

**Comment by MohammadHamdyGhanem** [comment]
Also, Replace method should return the No. of Replacements it did via an out param.
This code:
```
sb.Replace(x, y, out int c)
if (c ==0)
   MessageBox.Show("Not Found");
```

is more efficient that this one, which will search the sb twice in case the search string exists (suppose IndexOf added):
```
if (sb.IndexOf(x) == -1)
   MessageBox.Show("Not Found");
else
   sb.Replace(x, y)

```

**Comment by MohammadHamdyGhanem** [comment]
I made more suggestions about the StringBuilder.Replace method here #29498 and here #29499.

**Comment by MohammadHamdyGhanem** [comment]
I proposed to add a MutableString class that contains the missing StringBuilder functinality  #29510 
Hope you help implement it.

**Comment by dark2201** [comment]
That would be nice. In C++ `std::string` is mutable, so you can append using `+=` without performance concerns. In C# `+=` allocates new string every time. `StringBuilder` would be much better if it was named `MutableString` with additional string functionality but that would be a breaking change.

**Comment by jnm2** [comment]
I could really use `StringBuilder.IndexOf` today. Would it possibly be worth starting an issue that's less broad?

**Comment by GrabYourPitchforks** [comment]
Just as a note: you likely don't want to add all of the string functionality. For instance, APIs that are culture-aware (i.e., that take a `CultureInfo` or a `StringComparison`) only operate over fully contiguous buffers, _not_ discontiguous buffers like what `StringBuilder` is backed by today. Attempting to add these APIs to `StringBuilder` would almost certainly result in significant performance hits.

The better course of action IMO would be to open a separate issue detailing exactly which subset of APIs should be added to help immediately unblock you.
